src/main/java/org/broadinstitute/dsde/consent/ontology/datause/DataUseMatcherV3.java

```diff
@@ -0,0 +1,91 @@
+package org.broadinstitute.dsde.consent.ontology.datause;
+
+import com.google.inject.Inject;
+import org.apache.commons.lang3.BooleanUtils;
+import org.apache.commons.lang3.StringUtils;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.broadinstitute.dsde.consent.ontology.model.DataUseV3;
+import org.broadinstitute.dsde.consent.ontology.service.AutocompleteService;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+
+import static org.broadinstitute.dsde.consent.ontology.datause.DataUseMatchCasesV3.matchCommercial;
+import static org.broadinstitute.dsde.consent.ontology.datause.DataUseMatchCasesV3.matchDiseases;
+import static org.broadinstitute.dsde.consent.ontology.datause.DataUseMatchCasesV3.matchHMB;
+import static org.broadinstitute.dsde.consent.ontology.datause.DataUseMatchCasesV3.matchMDS;
+import static org.broadinstitute.dsde.consent.ontology.datause.DataUseMatchCasesV3.matchPOA;
+
+public class DataUseMatcherV3 {
+
+  private AutocompleteService autocompleteService;
+
+  public DataUseMatcherV3() {
+  }
+
+  @Inject
+  public void setAutocompleteService(AutocompleteService autocompleteService) {
+    this.autocompleteService = autocompleteService;
+  }
+
+  // Matching Algorithm
+  public ImmutablePair<Boolean, List<String>> matchPurposeAndDatasetV3(DataUseV3 purpose, DataUseV3 dataset) {
+    Map<String, List<String>> purposeDiseaseIdMap;
+    try {
+      purposeDiseaseIdMap = generatePurposeDiseaseIdMap(purpose.getDiseaseRestrictions());
+    } catch (Exception e) {
+      String purposeRestrictions = StringUtils.join(purpose.getDiseaseRestrictions(), ", ");
+      List<String> errors = Arrays.asList(e.getMessage(), "Error found in one of the purpose terms: " + purposeRestrictions);
+      return ImmutablePair.of(false, errors);
+    }
+
+    ImmutablePair<Boolean, List<String>> diseaseMatch = matchDiseases(purpose, dataset, purposeDiseaseIdMap);
+    final List<ImmutablePair<Boolean, List<String>>> matchReasons = new ArrayList<>();
+    matchReasons.add(matchDiseases(purpose, dataset, purposeDiseaseIdMap));
+    matchReasons.add(matchHMB(purpose, dataset));
+    matchReasons.add(matchMDS(purpose, dataset, diseaseMatch.getLeft()));
+    matchReasons.add(matchPOA(purpose, dataset));
+    matchReasons.add(matchCommercial(purpose, dataset));
```

Comment:
organization nit: Something that would be helpful would be to try and align the order here with the doc so it's easier to follow along and ensure that we're covering all cases. Additionally, we should probably re-structure the doc so we're doing the checks in a more consistent order. I would suggest doing the primary cases first (gru, hmb, disease) then the secondary cases after that.

Reply:
I completely agree, thanks!